javafmllanguage/src/main/java/net/minecraftforge/fml/javafmlmod/FMLModContainer.java

```diff
@@ -47,14 +52,79 @@ public FMLModContainer(IModInfo info, String className, ModFileScanData modFileS
         } catch (Throwable e) {
             LOGGER.error(LOADING, "Failed to load class {}", className, e);
             throw new ModLoadingException(info, ModLoadingStage.CONSTRUCT, "fml.modloading.failedtoloadmodclass", e);
         }
     }
 
+    /**
+     * <pre>
+     * Reads the Add-Exports and Add-Opens attributes (see note) from a mod file and
+     * attempts to apply them. This differers from the JEP in two significant ways:
+     *   1) Instead of opening things to ALL-UNNAMED it instead only opens to the
+     *      mod's module itself. Because I don't want other mods accidently relying
+     *      on transitive behavior
+     *   2) It is read from all mods not just the executable jars.
+     *
+     * From <a href = "https://openjdk.org/jeps/261">JEP 261: Module System</a>
+     *
+     * Two new JDK-specific JAR-file manifest attributes are defined to correspond
+     * to the --add-exports and --add-opens command-line options:
+     *     Add-Exports: &lt;module&gt;/&lt;package&gt;( &lt;module&gt;/&lt;package&gt;)*
+     *     Add-Opens: &lt;module&gt;/&lt;package&gt;( &lt;module&gt;/&lt;package&gt;)*
+     *
+     * The value of each attribute is a space-separated list of slash-separated
+     * module-name/package-name pairs. A &gt;module&lt;/&gt;package&lt; pair in
+     * the value of an Add-Exports attribute has the same meaning as the
+     * command-line option --add-exports &gt;module&lt;/&gt;package&lt;=ALL-UNNAMED.
+     *
+     * A &gt;module&lt;/&gt;package&lt; pair in the value of an Add-Opens attribute has the
+     * same meaning as the command-line option --add-opens &gt;module&lt;/&gt;package&lt;=ALL-UNNAMED.
+     *
+     * Each attribute can occur at most once, in the main section of a MANIFEST.MF file.
+     * A particular pair can be listed more than once. If a specified module was not
+     * resolved, or if a specified package does not exist, then the corresponding pair
+     * is ignored.
+     */
+    private void openModules(ModuleLayer layer, Module self, SecureJar jar) throws NoSuchMethodException, SecurityException, IllegalAccessException, InvocationTargetException {
+        var manifest = jar.moduleDataProvider().getManifest();
+        var entry = manifest.getMainAttributes().getValue("Add-Opens");
+        if (entry != null)
+            addOpenOrExports(layer, self, true, entry);
+        entry = manifest.getMainAttributes().getValue("Add-Exports");
+        if (entry != null)
+            addOpenOrExports(layer, self, false, entry);
+    }
+
+    private void addOpenOrExports(ModuleLayer layer, Module self, boolean open, String entry) throws NoSuchMethodException, SecurityException, IllegalAccessException, InvocationTargetException {
+        for (var pair : entry.split(" ")) {
+            var pts = pair.trim().split("/");
+            if (pts.length == 2) {
```

Comment:
I did this explicitly because it is required by the spec. And is how the [JVM itself](https://github.com/openjdk/jdk/blob/8b4749713c63a08e502845ed5d0a0236822018cd/src/java.base/share/classes/sun/launcher/LauncherHelper.java#L669) implements this method. I can add a branch to log a warning, but I wouldn't throw an exception.